File: 2075.cpp
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    string decodeCiphertext(string encodedText, int rows) {
        if (encodedText.empty()) return "";

        int n = encodedText.size();
        int cols = n / rows;
        string result = "";

        for (int startCol = 0; startCol < cols; startCol++) {
            int i = 0, j = startCol;

            while (i < rows && j < cols) {
                result += encodedText[i * cols + j];
                i++;
                j++;
            }
        }

        // Remove trailing spaces
        while (!result.empty() && result.back() == ' ') {
            result.pop_back();
        }

        return result;
    }
};

// Time: O(n)
// Space: O(n)

File: 2075.py
class Solution:
    def decodeCiphertext(self, encodedText: str, rows: int) -> str:
        if not encodedText:
            return ""
        
        cols = len(encodedText) // rows
        result = []
        
        for start_col in range(cols):
            i, j = 0, start_col
            while i < rows and j < cols:
                result.append(encodedText[i * cols + j])
                i += 1
                j += 1
        
        return "".join(result).rstrip()